### PR TITLE
Enable useSelect autocompletion

### PR DESCRIPTION
**Pull request by adamziel**
## What?

This PR builds `@wordpress/data` types to provide autocompletion support for `useSelect` across the repository.

The two included cosmetic changes in the `block-editor` package are here to showcase how it works:

https://user-images.githubusercontent.com/205419/175293659-4c009b3a-20f0-47e8-b358-0275204d114b.mp4

Note: For now, this only works with the store defined in the same package where `useSelect` is called. With this PR, TypeScript knows how to resolve the type of the `useSelect` hook, but all the cross-package store imports still come across as `any`.

## How

This PR adds a `wordpress__data.d.ts` ambient module that re-exports the types built for the `@wordpress/data` package.

The intention is to keep this change contained to the Gutenberg repository until the types mature more.

Once that happens, the next step would be publishing them by adding `"types": "build-types"` to `package.json` in `@wordpress/data`. It tells TypeScript how to resolve the typing file for `@wordpress/data` ([source](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html))

## Why?

Now that [useSelect is typed in an autocompletion-friendly way](https://github.com/WordPress/gutenberg/pull/41596), it only makes sense to enable that feature across the Gutenberg repository.

## Testing Instructions

Go to `use-available-alignments.js` and confirm the autocompletion works as on the video above.

CC @sarayourfriend @dmsnell @sirreal @jsnajdr @scruffian

### CODE REVIEW

**Comment by kevin940726** [comment]
Awesome work! This is probably my most wanted feature when working with `useSelect`. 💯

> but all the cross-package store imports still come across as `any`.

Any idea how we could solve this in the future?

> Unfortunately, this flag has multiple effects:
> 
> 1. It tells TypeScript how to resolve the typing file for `@wordpress/data` ([source](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html))
> 2. Includes the `build-types` directory in the published npm package ([source](https://github.com/WordPress/gutenberg/blob/20cf7e849fccc58b3fc7c84376033ba7e25a14c6/packages/README.md?plain=1#L285))
> 
> We want the former but not the latter. Before merging this PR we may need to decouple these behaviors. Perhaps there's a way to tell Lerna to ignore types for this package?

Why wouldn't we want the `build-types` directory to be included in the published npm package? I thought the [directory](https://unpkg.com/browse/@wordpress/data@6.11.0/build-types/) is already on the published package?

**Comment by adamziel** [comment]
@kevin940726 I somehow thought they're not being released – great spot! I am going to update the description to reflect that.

The reason I wanted to avoid releasing was to test-drive these types internally and make any necessary adjustments before releasing them to the world.

Since the build already includes types, adding `"types": "build-modules"` only tells TypeScript how to resolve the typing file for `@wordpress/data`. We could keep these types released but "unplugged" by having the `"types": "build-modules"` line  inside of this repository but not in the released packages. A small webpack plugin could get us there, I'm thinking about something that would see the following lines in package.json:

```js
{
    "types": "build-modules",
    "removeTypesResolutionOnBuild": true
}
```

And remove the `"types": "build-modules"` line from package.json.

> > but all the cross-package store imports still come across as any.
> Any idea how we could solve this in the future?

We'd have to add `tsconfig.json` and `"types": "build-modules"` to each of these packages. We'll get there step by step.

**Comment by adamziel** [comment]
Actually, a webpack plugin won't cut it as the package.json is bundled straight from `packages/data/package.json` and not from any intermediate directory with build artifacts. I can only think of one way of removing the `"types": "build-types"` on build: Adding an npm `postpack` hook that would rewrite the package.json inside of the final `.tgz` bundle.

**Comment by kevin940726** [comment]
> The reason I wanted to avoid releasing was to test-drive these types internally and make any necessary adjustments before releasing them to the world.

Well then, maybe we could take this another way around. Don't add the `types` field in `package.json`. Instead, referencing the types of the package to the directory at the project level. For instance, add a `@wordpress/core-data.d.ts` file in the `typings` folder and re-export all of the types in `build-types`. No idea if that would work though 😅

IMO, I don't think we have to go all the way just to _experiment_ with the types. For people not using TypeScript, the change doesn't matter to them. For people using TypeScript, they probably already have a custom config to ignore the type of the package anyway. I'd prefer just releasing the types in a minor release as a feature 😅. c.c. @gziolo for other concerns I missed though.

**Comment by adamziel** [comment]
> For people using TypeScript, they probably already have a custom config to ignore the type of the package anyway

I'd love that to be the case, but how do you know?

Also, I just realized this would be a breaking change as right now the typescript reads the types from the DefinitelyTyped repo whereas after this PR it would start reading them from the data package. The culprit is the two are not compatible :( 

Also interested in your thoughts @dmsnell and @sarayourfriend

**Comment by adamziel** [comment]
Unfortunately, updating the tarball won't work. Lerna creates a temporary file with a random name and does not pass its path to `postpack`:

https://github.com/lerna/lerna/blob/8ca18bedecf4f141c6242a099086e84b2ced72de/utils/pack-directory/lib/pack-directory.js#L63-L70

Which means we can do either of:

* Publish the npm packages with the `types` key in the package.json, as @kevin940726 suggested
* Experiment with a `@wordpress/core-data.d.ts` file, also as @kevin940726 suggested
* Remove the `"types"` from package.json before the build and restore it after the build. It sounds fragile, though.
* Rework the entire build pipeline. Let's just say I'd rather not.

I'd love to hear more opinions here.

**Comment by adamziel** [comment]
@kevin940726 I ended up going through with the ambient module declaration to enable this just in the Gutenberg repository for now. Would you be up for a re-review? Also cc @sirbrillig

File: typings/wordpress__data.d.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+declare module '@wordpress/data' {
+	export * from '@wordpress/data/build-types/index';
```
Comment (dmsnell) [inline typings/wordpress__data.d.ts:2]: how does this play with the issue of having a global DefinitelyTyped package in for `@wordpress/data`?

Comment (adamziel) [reply typings/wordpress__data.d.ts:2]: This overrides the DefinitelyTyped types thanks to the `"typeRoots": [ "./typings", "./node_modules/@types" ],` setting in tsconfig.json.

**Review by dmsnell — APPROVED** [review]
Let's see how this goes.

**Pull request merged**